src/Features/Core/Portable/Diagnostics/WorkspaceAnalyzerOptions.cs

```diff
@@ -10,13 +14,35 @@ namespace Microsoft.CodeAnalysis.Diagnostics
     /// </summary>
     internal sealed class WorkspaceAnalyzerOptions : AnalyzerOptions
     {
-        public WorkspaceAnalyzerOptions(AnalyzerOptions options, Workspace workspace)
+        private readonly Workspace _workspace;
+        private readonly OptionSet _optionSet;
+
+        public WorkspaceAnalyzerOptions(AnalyzerOptions options, OptionSet optionSet, Workspace workspace)
             : base(options.AdditionalFiles)
         {
-            this.Workspace = workspace;
+            _workspace = workspace;
+            _optionSet = optionSet;
         }
 
-        public Workspace Workspace { get; }
+        public HostWorkspaceServices Services => _workspace.Services;
+
+        public async Task<OptionSet> GetDocumentOptionSetAsync(SyntaxTree syntaxTree, CancellationToken cancellationToken)
+        {
+            var documentId = _workspace.CurrentSolution.GetDocumentId(syntaxTree);
```

Comment:
i don't like this.  why are we checking against CurrentSolution?   This seems like a race-nightmare.  Why not just grab the Solution snapshot when we make this object.  That way it is always valid as we're doing analysis, and we can always use it to map from a SyntaxTree to a document it?